[ADD] partner_firstname for v8

eLBati opened the pull request:
because it's working as is

eLBati committed:
[ADD] partner_firstname for v8
because it's working as is

eLBati committed:
[FIX] IntegrityError: column "lastname" contains null values
when creating or copying users

eLBati commented:
Fixed the error occurring when creating or copying users

:white_check_mark:

pedrobaeza commented:
You have a false positive on Travis. This is the error:

```
2014-07-23 22:14:24,970 19346 ERROR openerp_test openerp.sql_db: bad query: ALTER TABLE "res_partner" ALTER COLUMN "lastname" SET NOT NULL
Traceback (most recent call last):
File "/home/travis/build/OCA/odoo-8.0/openerp/sql_db.py", line 230, in execute
res = self._obj.execute(query, params)
IntegrityError: column "lastname" contains null values
2014-07-23 22:14:24,970 19346 WARNING openerp_test openerp.models: WARNING: unable to set column lastname of table res_partner not null !
Try to re-run: openerp-server --update=module
If it doesn't work, update records and execute manually:
ALTER TABLE res_partner ALTER COLUMN lastname SET NOT NULL
Traceback (most recent call last):
File "/home/travis/build/OCA/odoo-8.0/openerp/models.py", line 2664, in _auto_init
cr.execute('ALTER TABLE "%s" ALTER COLUMN "%s" SET NOT NULL' % (self._table, k))
File "/home/travis/build/OCA/odoo-8.0/openerp/sql_db.py", line 155, in wrapper
return f(self, *args, **kwargs)
File "/home/travis/build/OCA/odoo-8.0/openerp/sql_db.py", line 230, in execute
res = self._obj.execute(query, params)
IntegrityError: column "lastname" contains null values
```

To fix it, I think there must be a routine on init that fills _lastname_ with _name_ value.

Regards.

bwrsandman commented:
@pedrobaeza: this is exactly what [this function](https://github.com/eLBati/partner-contact/blob/8.0/partner_firstname/partner.py#L29) does. Maybe the API has changed the init functions.

Either way, it would be nice to have a logger tell us that the lastname field has been filled.

pedrobaeza commented:
We have to find then the problem. I wouldn't merge the module until we have this fixed, because it can cause a broken DB, don't you think?

Regards.

eLBati committed:
[IMP] partner_firstname - logging

eLBati committed:
[FIX] PEP8

eLBati commented:
```
Traceback (most recent call last):
File "/home/travis/build/OCA/odoo-8.0/openerp/models.py", line 2649, in _auto_init
cr.execute('ALTER TABLE "%s" ALTER COLUMN "%s" SET NOT NULL' % (self._table, k))
File "/home/travis/build/OCA/odoo-8.0/openerp/sql_db.py", line 155, in wrapper
return f(self, *args, **kwargs)
File "/home/travis/build/OCA/odoo-8.0/openerp/sql_db.py", line 230, in execute
res = self._obj.execute(query, params)
IntegrityError: column "lastname" contains null values
```

This happens because the `_auto_init` runs before the `init` and tries to set `required=True` for `lastname` field.
Just after that, the `init` correctly sets the NOT NULL constraint:

```
2014-07-26 00:10:48,523 19332 INFO openerp_test openerp.addons.partner_firstname.partner: NOT NULL constraint for res_partner.lastname correctly set
```

bwrsandman commented:
can `_auto_init` be overriden?
Is this an odoo 8.0 bug?

bwrsandman commented:
> Coverage increased (+8.44%)

I love seeing that! :smile:

eLBati commented:
> can _auto_init be overriden?

It should be overridable, but it does a lot of things and does not have hooks. It creates tables and adds constraints, so overriding it would not solve the problem.

Anyway I don't see a big problem with this error.
The purpose of `init` is to make sure the NOT NULL constraint on `lastname` to be set.
It tries with `'lastname': fields.char("Lastname", required=True)` (_auto_init) and if it fails, it adds the constraint with `init`. As the added log reports.
After having added the constraint, the error won't appear anymore.

bwrsandman commented:
I am sorry but, I cannot approve this while there are errors in the logs.

rvalyi commented:
"It should be overridable, but it does a lot of things and does not have hooks"
-> then may be it's time to propose a small refactor in official Odoo and lobby to get that in v8 while we may still have a few days to do so...

nhomar commented:
2014-07-26 10:17 GMT-04:30 Raphaël Valyi notifications@github.com:

> -> then may be it's time to propose a small refactor in official Odoo and
> lobby to get that in v8 while we may still have a few days to do so...

Absolutly agreed with this point.

## 

---

Saludos Cordiales

Nhomar G. Hernandez M.
+58-414-4110269
Skype: nhomar00
Web-Blog: http://geronimo.com.ve
Servicios IT: http://vauxoo.com
Linux-Counter: 467724
Correos:
nhomar@openerp.com.ve
nhomar@vauxoo.com
twitter @nhomar

hbrunn commented:
how about 'lastname': fields.char(...., oldname='name')?

eLBati commented:
> how about 'lastname': fields.char(...., oldname='name')?

I get

```
2014-07-29 10:40:50,067 25164 INFO demo_8 openerp.modules.module: module partner_firstname: creating or updating database tables
2014-07-29 10:40:50,160 25164 ERROR demo_8 openerp.sql_db: Programming error: column "name" of relation "res_partner" does not exist
, in query ALTER TABLE "res_partner" ALTER COLUMN "name" DROP NOT NULL
2014-07-29 10:40:50,160 25164 CRITICAL demo_8 openerp.service.server: Failed to initialize database `demo_8`.
Traceback (most recent call last):
  File "/home/elbati/workspace/odoo/buildout_cfg-odoo-8/parts/odoo/openerp/service/server.py", line 901, in preload_registries
    registry = RegistryManager.new(dbname, update_module=update_module)
  File "/home/elbati/workspace/odoo/buildout_cfg-odoo-8/parts/odoo/openerp/modules/registry.py", line 322, in new
    openerp.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/home/elbati/workspace/odoo/buildout_cfg-odoo-8/parts/odoo/openerp/modules/loading.py", line 357, in load_modules
    loaded_modules, update_module)
  File "/home/elbati/workspace/odoo/buildout_cfg-odoo-8/parts/odoo/openerp/modules/loading.py", line 263, in load_marked_modules
    loaded, processed = load_module_graph(cr, graph, progressdict, report=report, skip_modules=loaded_modules, perform_checks=perform_checks)
  File "/home/elbati/workspace/odoo/buildout_cfg-odoo-8/parts/odoo/openerp/modules/loading.py", line 163, in load_module_graph
    init_module_models(cr, package.name, models)
  File "/home/elbati/workspace/odoo/buildout_cfg-odoo-8/parts/odoo/openerp/modules/module.py", line 276, in init_module_models
    result = obj._auto_init(cr, {'module': module_name})
  File "/home/elbati/workspace/odoo/buildout_cfg-odoo-8/parts/odoo/openerp/api.py", line 204, in wrapper
    return old_api(self, *args, **kwargs)
  File "/home/elbati/workspace/odoo/buildout_cfg-odoo-8/parts/odoo/openerp/models.py", line 2582, in _auto_init
    cr.execute('ALTER TABLE "%s" ALTER COLUMN "%s" DROP NOT NULL' % (self._table, k))
  File "/home/elbati/workspace/odoo/buildout_cfg-odoo-8/parts/odoo/openerp/sql_db.py", line 155, in wrapper
    return f(self, *args, **kwargs)
  File "/home/elbati/workspace/odoo/buildout_cfg-odoo-8/parts/odoo/openerp/sql_db.py", line 230, in execute
    res = self._obj.execute(query, params)
ProgrammingError: column "name" of relation "res_partner" does not exist
```

eLBati committed:
[FIX] IntegrityError: column "lastname" contains null values

eLBati committed:
[FIX] partner_firstname - some PEP8

eLBati committed:
[FIX] partner_firstname - PEP8

eLBati committed:
[FIX] base_location PEP8

eLBati committed:
[FIX]  passport PEP8

eLBati committed:
[FIX] base_partner_merge PEP8

eLBati commented:
I fixed the error overriding `_set_default_value_on_column` instead of `init`: https://github.com/eLBati/partner-contact/commit/ae4c47d9edda38b46d827c28dd8cd6e2f9976ffe

I also fixed PEP8 for the whole repo, but travis doesn't collaborate and doesn't rebuild

pedrobaeza commented on partner_firstname/partner.py:
```diff
@@ -26,14 +30,23 @@ class ResPartner(orm.Model):
+            cr.execute('SELECT id FROM res_partner WHERE lastname IS NOT NULL '
+                       'Limit 1')
+            if not cr.fetchone():
+                cr.execute('UPDATE res_partner set lastname = name WHERE name '
```
Why don't you do it always? It will affect 0 or n rows, but it's as fast as making a select.

pedrobaeza commented on partner_firstname/partner.py:
```diff
@@ -26,14 +30,23 @@ class ResPartner(orm.Model):
+                # Create Sql constraint if table is not empty
+                cr.execute('SELECT id FROM res_partner Limit 1')
+                if cr.fetchone():
+                    cr.execute('ALTER TABLE res_partner ALTER COLUMN lastname '
```
Do it also always, having or not records. Maybe you can check if the restriction is already applied, but this way, you will always get message when there are records.

eLBati replied:
On 07/29/2014 04:42 PM, Pedro M. Baeza wrote:

> Do it also always, having or not records. Maybe you can check if the 
> restriction is already applied, but this way, you will always get 
> message when there are records.

Just ported from v7 version of the module, but you are right.
Modified it and used the query

```
UPDATE res_partner SET lastname = name WHERE name IS NOT NULL AND 
lastname IS NULL
```

to be sure to never overwrite lastname

eLBati committed:
[IMP] _set_default_value_on_column avoiding useless queries

pedrobaeza commented:
Please force a Travis rebuild making this in local repository in the updated PR branch:

```
git commit --ammend
git push <name_of_the_remote> <name_of_the_PR_branch> -f
```

Regards.

bwrsandman commented:
@pedrobaeza I can restart the build without having to do any amending

Done

pedrobaeza commented:
OK, as you want. It was only a trick I have discovered, because I don't have on some branches the possibility to make a rebuild on Travis.

eLBati commented:
Travis connection error is still there

pedrobaeza commented:
I don't know why is this happening. I haven't seen this error before in any repository.

eLBati commented:
On 07/30/2014 09:31 AM, Pedro M. Baeza wrote:

> I don't know why is this happening. I haven't seen this error before 
> in any repository.

It should be a temporary connection problem while downloading a PIP package

bwrsandman commented:
Restarted build

bwrsandman commented:
Still errored from `pip install -r /home/travis/maintainer-quality-tools/travis/requirements.txt`

eLBati commented:
I see green at https://travis-ci.org/OCA/partner-contact/builds/31171276

bwrsandman commented:
~~At~~
~~https://travis-ci.org/OCA/partner-contact/builds/31171276#L327~~
~~I see~~

```
error: [Errno 104] Connection reset by peer

Storing debug log for failure in /home/travis/.pip/pip.log

The command "travis_install_nightly ${VERSION}" failed and exited with 2 during .
```

Browse cache error, sorry.

bwrsandman commented:
All is well.
Thank you for your patience.
:+1:

bwrsandman commented:
You sure you don't want to convert it to the new api?

https://github.com/nbessi/odoo_new_api_guideline

pedrobaeza commented:
:+1: 

Thanks.

nbessi commented:
@bwrsandman I would prefer it too as it is quite a small addons, But nothing was formely decide by OCA. I have allready made this work during partner day work shop.

https://code.launchpad.net/~camptocamp/partner-contact-management/partner-first-name-new-api-v8

I should port this MP to github now that view and model are mostly frozen.

Regards

Nicolas

bwrsandman commented:
I have a feeling that if we don't port while converting to v8 (which we often do minimalistically), we will only have modules in the new API from people who feel like doing it.

@nbessi Can your openerp-conventions be made to test for the new API this on v8s?

lmignon commented on partner_firstname/partner.py:
```diff
@@ -26,14 +30,17 @@ class ResPartner(orm.Model):
+            _logger.info("NOT NULL constraint for "
+                         "res_partner.lastname correctly set")
+        return res
 
```
Why to not use install hooks? 
https://github.com/odoo/odoo/commit/4105b5f0285821d8056c721b5fd50b9cae902988

eLBati replied:
> Why to not use install hooks?
> odoo/odoo@4105b5f 
> https://github.com/odoo/odoo/commit/4105b5f0285821d8056c721b5fd50b9cae902988

How?

Anyway, I think `_set_default_value_on_column` is the right place to 
do it, as we are doing just that: setting a default value for the new 
column `lastname` and adding the NOT NULL constraint on it

lmignon replied:
After reading the code in openerp.models, you're right, the set_default_value_on_column is the right way to do it since it's only called on  the first installation or when upgrading the module 'ONLY' if the 'Not null' constrains is not yet applied. In proposing the 'install hooks', I wanted to avoid that this operation (_set_default_value_on_calumn) is called each time the module is upgraded.

eLBati commented:
On 07/30/2014 04:50 PM, Sandy wrote:

> You sure you don't want to convert it to the new api?
> 
> https://github.com/nbessi/odoo_new_api_guideline

I'm sure it would be good :smirk:

I just have no time in these days.

bwrsandman commented:
I tired it. It is not an easy port...

Coverting create to `@api.one` means the api will drop the `'name'` value for res users when creating related partner.

However, using the new fields, especially the compute fields are really good.

nbessi commented:
@bwrsandman I'm working on convention on my free time.
I have made major refactoring to support both API version and have all the correct bases to add required checks. I'm working on redefining name space correctly and intergrate stuff with travis. But I do not expect a release before the end of august.

bwrsandman commented:
We have two approves so far. Are we ready to merge or are there still objections?

JonathanNEMRY commented:
:+1:

hbrunn commented:
:+1:  - using the new api shouldn't be a prerequisite for porting currently, maybe at a later point.
